client/src/JobPaginatorState.js

```diff
@@ -0,0 +1,305 @@
+                ` go back to the main menu and hit the "build" button` +
+                ` for this job before publishing it.`
+        );
+        return false;
+    }
+    return !window.confirm(
+        "Are you sure you wish to publish this job? You will not be able to make any further edits."
+    );
+};
+
+const getLabel = (order, imageNumber) => {
+    return typeof order[imageNumber] === "undefined" ? null : order[imageNumber]["label"];
+};
+
+const countMagicLabels = (order, startAt) => {
+    let count = 0;
+    for (let startIndex = startAt; startIndex < order.length; startIndex++) {
+        if (!getLabel(order, startIndex)) {
+            count++;
+        }
+    }
+    return count;
+};
+
+const validateSave = (order) => {
+    const count = countMagicLabels(order, 0);
+    return count > 0 && !confirmSavedMagicLabels(count);
+};
+
+const useJobPaginator = (initialCategory, initialJob) => {
+    const [active, setActive] = useState(false);
+    const [currentPage, setCurrentPage] = useState(0);
+    const [zoom, setZoom] = useState(false);
+    const [order, setOrder] = useState([]);
+    const [magicLabelCache, setMagicLabelCache] = useState([]);
+    const [category, setCategory] = useState(initialCategory);
+    const [job, setJob] = useState(initialJob);
+
+    const saveMagicLabels = () => {
+        order.forEach((o, orderIndex) => {
+            if (!getLabel(order, orderIndex)) {
+                setLabel(orderIndex, getMagicLabel(orderIndex));
+            }
+        });
+    };
+
+    const setLabel = (imageNumber, text) => {
+        setMagicLabelCache([]); // clear label cache whenever there is a change
+        if (getLabel(order, imageNumber)) {
+            return;
+        }
+        order[imageNumber]["label"] = text !== null && text.length === 0 ? null : text;
+        setOrder(order);
+        dispatchEvent(new Event("Prep.editted"));
+    };
+
+    const getMagicLabel = (imageNumber) => {
+        const label = getLabel(order, imageNumber);
+
+        if (!label) {
+            return label;
+        }
+        if (typeof magicLabelCache[imageNumber] === "undefined") {
+            magicLabelCache[imageNumber] = MagicLabeler.getLabel(imageNumber, getMagicLabel);
+            setMagicLabelCache(magicLabelCache);
+        }
+        return magicLabelCache[imageNumber];
+    };
+
+    const setPage = (page) => {
+        if (page >= 0 && page < order.length) {
+            setCurrentPage(page);
+        }
+    };
+
+    const nextPage = () => {
+        setPage(currentPage + 1);
+    };
+
+    const prevPage = () => {
+        setPage(currentPage - 1);
+    };
+
+    const updatePagesByStatus = (status) => {
+        const {
+            file_problems: { deleted, added },
+        } = status;
+        const message = [];
+        if (deleted.length > 0) {
+            message.push(deleted.length + " file(s) have been removed = require(the job since the last edit.\n)");
```

Comment:
This should be "have been removed from the job" -- looks like perhaps a search-and-replace or other IDE feature malfunctioned in the past. (I realize that this problem was not introduced by this PR, but we might as well fix it while we see it). Here's the original code from the previous iteration of the project: https://github.com/FalveyLibraryTechnology/VuDLPrep/blob/f59a7529aa4bb6afecd4508ca5284637a23ee3c9/app/assets/javascripts/components/JobPaginator.js.jsx#L120

Reply:
I also realized that the "\n" is in the wrong place in this string (something @crhallberg fixed in #71, but which hasn't been merged anywhere else).

Reply:
I made the change.